### PR TITLE
feat(ic-http-certification): add common constructs to skip certification

### PR DESCRIPTION
**Pull request by nathanosdev**
TT-425

- Adds some common code for skipping certification to the `ic-http-certification` library.
- Extends the custom asset canister example to show how to skip certification for certain paths, along with docs and an e2e test using PocketIC.
- Adds a new example to show how to skip certification for ALL responses, along with docs and an e2e test using PocketIC.

Inspired by this PR to skip certification for Candid UI: https://github.com/dfinity/candid/pull/571

### CODE REVIEW

File: examples/http-certification/custom-assets/README.md [hunk 1]
```diff
@@ -197,9 +208,6 @@ fn certify_asset_response(
                 asset_tree_path,
                 &certification,
             ));
-
-            // set the canister's certified data
-            set_certified_data(&http_tree.root_hash());
```
Comment (nathanosdev) [inline examples/http-certification/custom-assets/README.md:202]: This change is unrelated to the main change, I just noticed while I was here that we were updating the certified variable every time an asset was certified, but they're all certified on init and upgrade so we can just update this once after all files are certified.

File: examples/http-certification/custom-assets/README.md [hunk 2]
```diff
@@ -141,11 +142,11 @@ Certifying responses is more involved here compared to the simpler approach used
 The first step is defining a reusable function to create a response with all of the necessary default headers. This function is very similar to the counterpart in the [JSON API](https://internetcomputer.org/docs/current/developer-docs/http-compatible-canisters/serving-json-over-http) example with the biggest difference being in the headers that are used. Since the responses from an API serving static assets will be rendered directly in the browser, more security-focused headers are necessary:
 
 ```rust
-fn create_asset_response(
+fn get_asset_headers(
     additional_headers: Vec<HeaderField>,
     body: &[u8],
     cel_expr: String,
-) -> HttpResponse {
+) -> Vec<(String, String)> {
```
[the remark this answers is not in the document]
Comment (nathanosdev) [reply examples/http-certification/custom-assets/README.md:149]: Good point, thanks.

File: examples/http-certification/custom-assets/README.md [hunk 3]
```diff
@@ -383,22 +388,56 @@ fn certify_index_asset() {
 }
 ```
 
+It's also possible to skip certification for certain routes. This can be relatively easily as follows:
```
[the remark this answers is not in the document]
Comment (nathanosdev) [reply examples/http-certification/custom-assets/README.md:389]: Thanks!

File: examples/http-certification/custom-assets/README.md [hunk 4]
```diff
@@ -383,22 +388,56 @@ fn certify_index_asset() {
...
+
+```rust
+const DYNAMIC_REQ_PATH: &str = "/dynamic";
```
[the remark this answers is not in the document]
Comment (nathanosdev) [reply examples/http-certification/custom-assets/README.md:392]: Agreed, thanks

File: examples/http-certification/custom-assets/README.md [hunk 5]
```diff
@@ -383,22 +388,56 @@ fn certify_index_asset() {
...
+
+fn certify_dynamic_asset() {
```
[the remark this answers is not in the document]
Comment (nathanosdev) [reply examples/http-certification/custom-assets/README.md:394]: It's really super confusing, because actually we add a certification for this path that tells the gateway to skip verification for the response. So there is actually some verification steps involved to make sure that it is not skipped arbitrarily.

I used `add_certification_skips` instead, which I think is clearer, let me know if it's still confusing.

File: examples/http-certification/custom-assets/README.md [hunk 6]
```diff
@@ -383,22 +388,56 @@ fn certify_index_asset() {
...
+
+    update_certified_data();
 }
 ```
 
 ## Serving assets
 
 With all assets certified, they can be served over HTTP. The steps to follow when serving assets are:
 
+- Check if the request path matches the dynamic path.
```
[the remark this answers is not in the document]
Comment (nathanosdev) [reply examples/http-certification/custom-assets/README.md:437]: Thanks!

File: examples/http-certification/custom-assets/README.md [hunk 7]
```diff
@@ -490,6 +540,46 @@ fn asset_handler(req: &HttpRequest) -> HttpResponse<'static> {
 }
 ```
 
+Creating the dynamic response is done as follows:
```
[the remark this answers is not in the document]
Comment (nathanosdev) [reply examples/http-certification/custom-assets/README.md:541]: Thanks!

File: examples/http-certification/skip-certification/README.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+This is a relatively simple guide so there's no prerequisites as such, but it's recommended to check out the full certification guides to make sure that certification is not a good fit for your project.
+
+- [x] Complete the ["Serving static assets over HTTP"](https://internetcomputer.org/docs/current/developer-docs/web-apps/http-compatible-canisters/serving-static-assets-over-http) guide.
+- [x] Complete the ["Custom HTTP Canisters"](https://internetcomputer.org/docs/current/developer-docs/http-compatible-canisters/custom-http-canisters) guide.
+- [x] Complete the ["Serving JSON over HTTP"](https://internetcomputer.org/docs/current/developer-docs/http-compatible-canisters/serving-json-over-http) guide.
+
+# Skipping certification
+
+Skipping certification for all responses is a relatively simple task that can be completed in 2 very simple steps.
```
[the remark this answers is not in the document]
Comment (nathanosdev) [reply examples/http-certification/skip-certification/README.md:21]: Good catch, thanks

File: examples/http-certification/skip-certification/README.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+First, set the canister's certified data in the canister's `init` lifecycle hook:
+
+```rust
+#[init]
+fn init() {
+    set_certified_data(&skip_certification_certified_data());
+}
+```
```
[the remark this answers is not in the document]
Comment (nathanosdev) [reply examples/http-certification/skip-certification/README.md:30]: Thanks

File: examples/http-certification/skip-certification/README.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+```rust
+#[query]
+fn http_request() -> HttpResponse<'static> {
+    let mut response = create_response();
+
+    add_skip_certification_header(data_certificate().unwrap(), &mut response);
+
+    response
+}
+```
```
[the remark this answers is not in the document]
Comment (nathanosdev) [reply examples/http-certification/skip-certification/README.md:45]: Thanks

File: packages/ic-http-certification/src/utils/skip_certification.rs [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+///             "certificate=:AQID:, tree=:2dn3gwJJaHR0cF9leHBygwJDPCo+gwJYIMMautvQsFn51GT9bfTani3Ah659C0BGjTNyJtQTszcjggNA:, expr_path=:2dn3gmlodHRwX2V4cHJjPCo+:, version=2".to_string(),
+///         ),
+///         (
+///             CERTIFICATE_EXPRESSION_HEADER_NAME.to_string(),
+///             DefaultCelBuilder::skip_certification().to_string()
+///         ),
+///     ]
+/// );
+/// ```
+pub fn add_skip_certification_header(data_certificate: Vec<u8>, response: &mut HttpResponse) {
```
[the remark this answers is not in the document]
Comment (nathanosdev) [reply packages/ic-http-certification/src/utils/skip_certification.rs:47]: I added a simple test for `skip_certification_certified_data` to ensure the returned hash doesn't change unexpectedly. We know the current hash returned is correct because of the e2e tests on the example projects.

I think for `add_skip_certification_header` that the docs test should serve the purpose of the unit test, similar to the other PR. Do you think it should not be the case here?

**Review by przydatek — APPROVED** [review]
LGTM, thanks! 
(I haven't reviewed `.ts`-files though...)
